### PR TITLE
Administration: delete action

### PR DESCRIPTION
**Pull request by kpsherva**
Screenshot 2022-08-29 at 16 07 32

### CODE REVIEW

File: invenio_administration/admin.py [hunk 1]
```diff
@@ -103,7 +103,12 @@ def add_view(self, view, view_instance, *args, **kwargs):
             view_func=view,
         )
 
-        self._menu.add_view_to_menu(view_instance)
+        from invenio_administration.views.base import AdminResourceDetailView, \
+            AdminResourceEditView
+
+        if not isinstance(view_instance, AdminResourceDetailView) and not isinstance(
```
Comment (alejandromumo) [inline invenio_administration/admin.py:109]: Personally I prefer `allow` over `block`, I think it scales better. E.g. this condition will block `DetailView` and `EditView` from being added to the menu views. But every other view will be added. For now, that means only `ListView` is added.  However, in the future we might introduce another type of view and it will be a "bug". 

It's not a major issue because we do not foresse any other types of views for now, but we never know. 

```suggestion
        if isinstance(view_instance, AdminResourceListView):
```

File: invenio_administration/admin.py [hunk 2]
```diff
@@ -103,7 +103,12 @@ def add_view(self, view, view_instance, *args, **kwargs):
             view_func=view,
         )
 
-        self._menu.add_view_to_menu(view_instance)
+        from invenio_administration.views.base import AdminResourceDetailView, \
```
Comment (alejandromumo) [inline invenio_administration/admin.py:106]: imports are here for a reason?

File: invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      displayDelete,
+      actions,
+      apiEndpoint,
+      resource,
+      successCallback,
+      idKeyPath,
+    } = this.props;
+
+    // if number of actions is greater than 3, we display all in a dropdown
+    const displayAsDropdown =
```
Comment (alejandromumo) [inline invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js:18]: (from CI): this is unused by now, is that on purpose?

File: invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Delete.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      >
+        <Icon name="trash alternate" />
+        Delete
+      </Button>
+    );
+  }
+}
+
+Delete.propTypes = {
+  displayEdit: PropTypes.bool,
```
Comment (alejandromumo) [inline invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Delete.js:44]: is this prop needed here?

File: invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultItem.js [hunk 5]
```diff
@@ -23,12 +43,40 @@ export class SearchResultItem extends Component {
             </Table.Cell>
           );
         })}
+        {resourceHasActions && (
+          <Table.Cell>
+            <Actions
```
Comment (jennur) [inline invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultItem.js:48]: Should this not to be a modal? Right now the action will trigger immediately on the button-click, no?

File: invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+import { APIRoutes } from "./routes";
+import { http } from "./config";
+
+const getResource = async (
+  apiEndpoint = undefined,
```
Comment (alejandromumo) [inline invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js:7]: I think default values as `undefined` is the default behaviour of javascript.

File: invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  pid= undefined,
+) => {
+  return await http.get(APIRoutes.get(apiEndpoint, pid));
```
Comment (alejandromumo) [inline invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js:10]: Why is this function awaiting the result if it does not need it? Is it to have better debugging? Otherwise, `return await` is not necessary and has a tiny performance overhead.

see https://eslint.org/docs/latest/rules/no-return-await

File: invenio_administration/assets/semantic-ui/js/invenio_administration/api/config.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+// under the terms of the MIT License; see LICENSE file for more details.
+
+import axios from "axios";
+
+const apiConfig = {
+  withCredentials: true,
+  xsrfCookieName: "csrftoken",
+  xsrfHeaderName: "X-CSRFToken",
+  headers: {
+    Accept: "application/json",
```
Comment (alejandromumo) [inline invenio_administration/assets/semantic-ui/js/invenio_administration/api/config.js:14]: Tiny minor ocd comment (😄) : use double quotes as below just for consistency

```suggestion
    "Accept": "application/json",
```

File: invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultOptions.js [hunk 9]
```diff
@@ -33,7 +33,6 @@ export const SearchResultOptions = ({ sortOptions, sortOrderDisabled }) => {
               ariaLabel={i18next.t("Sort")}
               label={(cmp) => (
                 <>
```
Comment (alejandromumo) [inline invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultOptions.js:35]: are these fragments needed anymore?

Comment (alejandromumo) [inline invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultOptions.js:35]: are these fragments needed anymore?

File: invenio_administration/ext.py [hunk 10]
```diff
@@ -23,6 +23,7 @@ def __init__(self, app=None, entry_point_group="invenio_administration.views"):
         self.entry_point_group = entry_point_group
 
         self.administration = None
+        self.registered_extensions = []
```
Comment (alejandromumo) [inline invenio_administration/ext.py:26]: why is this needed inside the extension?

File: invenio_administration/views/base.py [hunk 11]
```diff
@@ -165,30 +159,47 @@ def _get_resource(cls, extension_name=None):
...
+             "order": 1
+             }
+         }
         """
+        serialized_actions = {}
+        for key, value in self.actions:
+            if "payload_schema" and "order" not in value:
+                raise InvalidActionsConfiguration
+
+            if value["payload_schema"] is not None:
```
Comment (alejandromumo) [inline invenio_administration/views/base.py:197]: `value["payload_schema"]` was already validated before, I think it's safe to access here

File: invenio_administration/views/base.py [hunk 12]
```diff
@@ -165,30 +159,47 @@ def _get_resource(cls, extension_name=None):
...
+                serialized_actions[key]["payload_schema"] = \
+                    self._schema_to_json(value["payload_schema"])
+            else:
+                value["payload_schema"] = None
```
Comment (alejandromumo) [inline invenio_administration/views/base.py:201]: I am not sure what is being doing here. Does admin allow for `payload_schema` to be `None`? If the action ends up not being serialised, due to lack of payload schema, why does it mutate the action object like this?

File: invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+};
+
+const deleteResource = async (
```
Comment (alejandromumo) [inline invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js:13]: Missing JSDoc for multiple functions and components in this PR.

**Comment by alejandromumo** [comment]
Thanks for the PR!

Can you add some screenshots?

File: invenio_administration/admin.py (hunk 1 above)
Comment (kpsherva) [reply invenio_administration/admin.py:109]: what about BaseView if someone wants to go custom? this is why I went with excluding rather than picking only the list views

File: invenio_administration/admin.py (hunk 2 above)
Comment (kpsherva) [reply invenio_administration/admin.py:106]: its because of circular imports

File: invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js (hunk 3 above)
Comment (kpsherva) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js:18]: yes, to be needed later

File: invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+      displayEdit && displayDelete && Object.keys(actions).length > 1;
+
+    return (
+      <Delete
```
[the remark this answers is not in the document]
Comment (kpsherva) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js:22]: I will create issues for missing pieces

File: invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js (hunk 6 above)
Comment (kpsherva) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js:7]: yes it is, just a habit of mine to have things explicit

File: invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js (hunk 7 above)
Comment (kpsherva) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js:10]: why it is not necessary ? it is an async call

File: invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+      actions,
+      apiEndpoint,
+      resource,
+      successCallback,
+      idKeyPath,
+    } = this.props;
+
+    // if number of actions is greater than 3, we display all in a dropdown
+    const displayAsDropdown =
+      displayEdit && displayDelete && Object.keys(actions).length > 1;
```
[the remark this answers is not in the document]
Comment (kpsherva) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js:19]: not implemented yet, to do

File: invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultItem.js (hunk 5 above)
Comment (kpsherva) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultItem.js:48]: I opened an issue to add a modal, I mentioned this on the standup

File: invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultsContainer.js [hunk 16]
```diff
@@ -20,6 +29,7 @@ export const SearchResultsContainer = ({ results, columns }) => {
               <Table.HeaderCell key={property + order}>{text}</Table.HeaderCell>
             );
           })}
+          {resourceHasActions && <Table.HeaderCell>Actions</Table.HeaderCell>}
```
[the remark this answers is not in the document]
Comment (kpsherva) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultsContainer.js:32]: thanks I will make a pass to mark all translations in the next PR

File: invenio_administration/ext.py (hunk 10 above)
Comment (kpsherva) [reply invenio_administration/ext.py:26]: leftover, thanks for spotting it

File: invenio_administration/views/base.py (hunk 12 above)
Comment (kpsherva) [reply invenio_administration/views/base.py:201]: we might have an action without a payload (empty POST) which we need to communicate to the frontend but you are right there is no need of assigning None explictly again

File: invenio_administration/admin.py (hunk 1 above)
Comment (alejandromumo) [reply invenio_administration/admin.py:109]: That's a strong point. 

Alternatively, in case someone wants to go full custom, we could provide a `CustomView` that inherits from `BaseView` just so we can have more control on views management. That would make `BaseView` only used internally.

File: invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js (hunk 3 above)
Comment (alejandromumo) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/actions/Actions.js:18]: would be nice to have a `TODO` there. Helped me a lot to visualize missing parts when the module was firstly created.

File: invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js (hunk 6 above)
Comment (alejandromumo) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js:7]: No worries then :) I am just not used to see that in javascript, that's why I asked

File: invenio_administration/admin.py (hunk 1 above)
Comment (kpsherva) [reply invenio_administration/admin.py:109]: Or we rename `BaseView` to just `AdminView`

File: invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultOptions.js (hunk 9 above)
Comment (kpsherva) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/search/SearchResultOptions.js:35]: yes otherwise react-searchkit complains

File: invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js (hunk 7 above)
Comment (alejandromumo) [reply invenio_administration/assets/semantic-ui/js/invenio_administration/api/actions.js:10]: it is unnecessary, as this function is not consuming the promise's return value or its error. It's just returning the promise back.  See the `eslint` rule explanation or this [explanation](https://maximorlov.com/linting-rules-for-asynchronous-code-in-javascript/#6-no-return-await)

In this case, note that you already `await` for the result of this promise when you actually need to: 

``` javascript
async function deleteEndpoint() {
    return await http.delete()  // this returns a promise nonetheless
}

async function myFunc(){
    try {
        const result = await deleteEndpoint(); // promise's return value is consumed here
        console.log(result);
    } catch (error) {
        console.log(error);  // promise's rejection is handled here
    }
}
```

if you don't stop the event loop here, you will get an `undefined` as it will move on and won't stop to actually resolve the promise returned by `deleteEndpoint`.

Also note that not all async calls needs to be awaited, sometimes the event loop does not need to wait for the result of a promise to move on. You can have some performance gains from this concurrency. But that's another, unrelated discussion. See this [example](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function#async_functions_and_execution_order) if you are curious about promises and concurrency.

File: invenio_administration/admin.py (hunk 1 above)
Comment (alejandromumo) [reply invenio_administration/admin.py:109]: We would benefit from having a more objective views management. E.g. every view is a `BaseView`, thus we can't differentiate a `CustomView` from a `ListView` if we want to operate only on one of those. 

It is not blocking, so we can also analyse this in another issue. Especially for this `allow` or `block` situation

**Review by jennur — APPROVED** [review]
